### PR TITLE
Handle directory-file and file-directory comparisons in the diff

### PR DESCRIPTION
**Pull request by TanmayPatil105**
GNU diff treats `diff DIRECTORY FILE` as `diff DIRECTORY/FILE FILE`

```console
$ echo a > a
$ mkdir d
$  echo da > d/a
$ diff -u d a
--- d/a	2024-04-20 22:16:11.145369745 +0530
+++ a	2024-04-20 22:15:57.785813179 +0530
@@ -1 +1 @@
-da
+a
$ diff -u a d
--- a	2024-04-20 22:15:57.785813179 +0530
+++ d/a	2024-04-20 22:16:11.145369745 +0530
@@ -1 +1 @@
-a
+da
$ cargo run -- -u d a
    Finished dev [unoptimized + debuginfo] target(s) in 0.02s
     Running `target/debug/diffutils -u d a`
--- d/a	2024-04-20 22:16:11.145369745 +0530
+++ a	2024-04-20 22:15:57.785813179 +0530
@@ -1 +1 @@
-da
+a
$ cargo run -- -u a d
    Finished dev [unoptimized + debuginfo] target(s) in 0.02s
     Running `target/debug/diffutils -u a d`
--- a	2024-04-20 22:15:57.785813179 +0530
+++ d/a	2024-04-20 22:16:11.145369745 +0530
@@ -1 +1 @@
-a
+da

```

### CODE REVIEW

**Comment by TanmayPatil105** [comment]
Well, this will handle `diff -u d a` but not `diff -u - a < d`.

```console
$ diff -u - a < d
--- -/a	2024-04-20 22:16:11.145369745 +0530
+++ a	2024-04-20 22:15:57.785813179 +0530
@@ -1 +1 @@
-da
+a
$ cargo run -- -u - a < d
    Finished dev [unoptimized + debuginfo] target(s) in 0.10s
     Running `target/debug/diffutils -u - a`
Failed to read from-file: Is a directory (os error 21)
```

GNU diff (3.9+) implementes `stdin` directory using `openat` 
See: https://git.savannah.gnu.org/cgit/diffutils.git/tree/src/diff.c#n1541

**Comment by sylvestre** [comment]
sorry but your test fails on windows with:

```
---- read_from_directory stdout ----
---- read_from_directory stderr ----
thread 'main' panicked at tests\integration.rs:257:5:
assertion `left == right` failed
  left: "--- target/integration/d\\a\tTIMESTAMP\n+++ target/integration/a\tTIMESTAMP\n@@ -1 +1 @@\n-da\n+a\n"
 right: "--- target/integration/d/a\tTIMESTAMP\n+++ target/integration/a\tTIMESTAMP\n@@ -1 +1 @@\n-da\n+a\n"
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace


```

**Comment by oSoMoN** [comment]
> Well, this will handle `diff -u d a` but not `diff -u - a < d`.

That would be nice to have, but it can be implemented in a followup PR.

File: tests/integration.rs [hunk 1]
```diff
@@ -234,3 +236,50 @@ fn read_from_stdin() -> Result<(), Box<dyn std::error::Error>> {
 
     Ok(())
 }
+
+#[test]
+fn read_from_directory() -> Result<(), Box<dyn std::error::Error>> {
```
Comment (oSoMoN) [inline tests/integration.rs:241]: Suggestion: rename the test "compare_file_to_directory".

File: tests/integration.rs [hunk 2]
```diff
@@ -234,3 +236,50 @@ fn read_from_stdin() -> Result<(), Box<dyn std::error::Error>> {
...
+    let target = PathBuf::from("target/integration");
+    let _ = std::fs::create_dir(&target);
```
Comment (oSoMoN) [inline tests/integration.rs:243]: This creates test directories in the current working dir, without checking first whether they already exist. I got bitten by that in #61. For extra robustness, can you do that in a temporary directory (see [tempfile::tempdir()](https://docs.rs/tempfile/3.10.1/tempfile/fn.tempdir.html)) to ensure pristine test data, and that it gets deleted at the end of the test?

**Pull request merged**